### ISSUE TITLE
Trained ImageNet model

### ISSUE THREAD

**Issue by erfannoury** [benanne/Lasagne#157]
Hello everyone.
Thanks for the great job.

I was wondering if there is a pre-trained Lasagne model on ImageNet available? I have searched almost everywhere but couldn't find a pre-trained model on Lasagne, Theano itself, or any other library dependent on Theano. 
On the other hand, can pre-trained models on other libraries, like Caffe and Torch get converted to Lasagne models?

**Comment by f0k** [comment]
Hey, thanks for your interest.

> I was wondering if there is a pre-trained Lasagne model on ImageNet available?

No, not yet. Lasagne hasn't been released yet, but when it is, chances are that someone will train it on ImageNet at some point.

> On the other hand, can pre-trained models on other libraries, like Caffe and Torch get converted to Lasagne models?

Lasagne does not provide methods for saving or loading a model. This is on on purpose; its scope is currently limited to creating Theano expressions for training and applying feed-forward neural networks. Model serialization has to be done by the user (with the help of `get_all_param_values()` or `pickle`, for example).
This being said, converting a model from Caffe or Torch would require a couple of things:

- Find a way to read the model in Python
- Ensure all required layer types are implemented in Lasagne
- Recreate the network in terms of Lasagne `Layer` instances
- Set the weight, bias, ... parameters of the `Layer` instances to the values read from the Caffe or Torch model

That's quite some work, but yes, in theory, it can be done. (Again, this is outside the scope of the main library.)

**Comment by erfannoury** [comment]
> No, not yet. Lasagne hasn't been released yet, but when it is, chances are that someone will train it on ImageNet at some point.

So I'm looking forward to Lasagne's first release. Hopefully then I can find a pre-trained model of ImageNet.

> Lasagne does not provide methods for saving or loading a model.

I got it. But aren't you even planning on implementing saving and loading of models in future? 

>That's quite some work, but yes, in theory, it can be done.

Well, that's quite a lot of work :smiley:, specially for beginners like me. 

So, it looks like I have got some learning to do.

**Comment by Britefury** [comment]
If you install joblib (`pip install joblib`) you can use its pickle-like saving and loading functionality just fine. Ensure that you retain a reference to the final layer and any input variables. For example, here is a class that could be used to represent an image classifier:

```
class LasagneImageClassifier (object):
    def __init__(self, input_layer, output_layer):
        self.input_layer = input_layer
        self.output_layer = output_layer

    @property
    def layers(self):
        return lasagne.layers.get_all_layers(self.output_layer)

    @property
    def batch_size(self):
        return self.input_layer.shape[0]

    @property
    def input_variable(self):
        return self.input_layer.input_var

    @property
    def predicted_class_probabilities_expr(self):
        return self.output_layer.get_output(deterministic=True)

    def predict_image_class_probabilities(self, images_as_tensor):
        # Get the batch size and input variable x
        batch_size = self.batch_size
        x = self.input_variable

        # Create a theano function for obtaining class probs
        get_class_probabilities = theano.function([x], self.predicted_class_probabilities_expr)

        # Invoke it
        return get_class_probabilities(images_as_tensor.astype(theano.config.floatX)))
```


To save it:
```
joblib.dump(classifier_instance, classifier_path)
```

To load it:
```
classifier_instance = joblib.load(classifier_path)
```

**Comment by eickenberg** [comment]
Also take a look at this wrapper we wrote for [`sklearn-theano`](http://sklearn-theano.github.io): https://github.com/sklearn-theano/sklearn-theano/blob/master/sklearn_theano/feature_extraction/caffe/caffemodel.py#L152. It gives you all the Theano expressions of all intermediate layers. We have a working feedforward example for [GoogleNet caffe model](https://github.com/sklearn-theano/sklearn-theano/blob/master/sklearn_theano/feature_extraction/caffe/googlenet.py), but it should be immediate or straightforward to use other models, too. These can also immediately interact with Lasagne.

As an aside: We are trying to promote sklearn-theano as a sort of bay for easily usable pretrained models, with emphasis on the fact that they be turned into sklearn-compatible estimators and use Theano. That way both users that only want to use and users that need theano expressions can profit from the architecture.

Since the class architecture in sklearn-theano is a rudimentary version of what Lasagne has developed, it may be interesting to evaluate whether there are deeper synergies to be exploited here?

CC @kastnerkyle

**Comment by erfannoury** [comment]
@Britefury Thanks for the tip. That will be of great use.

@eickenberg I checked it out. I think it's a great solution. But I can't test it, since I don't have caffe installed yet (actually I'm on Windows and it makes installing such libraries non-trivial). I will get around this problem and try it out. Thank you

**Comment by benanne** [comment]
@eickenberg absolutely! yay synergies! In the long term we'd love to have a way to wrap a simple Lasagne model definition into an object that supports the sklearn interface. This has always been the plan but we've been focusing on other things for now. Anyway, so as not to hijack this issue, if you want to discuss this further feel free to create a new issue, or maybe a post on the mailing list: https://groups.google.com/forum/#!forum/lasagne-users

**Comment by erfannoury** [comment]
I could finally run the test script for GoogLeNet using sklearn_theano. Thank you @eickenberg for the suggestion. 
I think this issue can be closed now. 

Thank you everyone for your help.

**Comment by eickenberg** [comment]
> This has always been the plan but we've been focusing on other things for now.

OK, cool, let's discuss this as soon as it crops up on the horizon. There are several ways one could imagine, but they will depend crucially on the states of both projects at that time. I will open an early issue when I get some time to brainstorm (unfortunately not immediately, I have a series of deadlines to sprint towards).

**Comment by willard-yuan** [comment]
@erfannoury You can view [visual-semantic-embedding](https://github.com/ryankiros/visual-semantic-embedding). ryankiros provided the pre-trained model `vgg19.pkl`, but he didn't provided the training code.

**Comment by kastnerkyle** [comment]
I think a bigger question is looming here - how can we share models between projects, even moreso between projects which are in different languages (e.g. Torch, Caffe, Theano)? I have talked with Fred about this some, but can say that heavily using/depending on protobuf seems like the "wrong thing" to me (even though Caffe is doing it, and we probably can't run away now :/ ). It has a lot of (IMO) terrible underlying problems, and seems to be solving many problems we as a community do not have. Maybe HDF5 with a known set of names/conversions for columns (kind of like Keras has done) is one way forward for python people at least?

With respect to sklearn-theano and Lasagne compatibility, I would be more than happy to adopt Lasagne style layers as long as we can still confirm numerical accuracy with respect to the original networks. We should definitely look at this.

**Comment by kastnerkyle** [comment]
Also, we have another outstanding PR which uses GoogLeNet to at least classify ImageNet. No finetuning or anything, ut might be of help for people who are interested. https://github.com/sklearn-theano/sklearn-theano/pull/85

**Comment by ebenolson** [comment]
We've had a [Model Zoo](https://github.com/Lasagne/Recipes/tree/master/modelzoo) in Lasagne/Recipes for a while now, containing versions of GoogLeNet and several VGG models from Caffe's Model Zoo. 

But a better standard for sharing models, or tools for automatically converting between languages would definitely be cool..

**Comment by erfannoury** [comment]
@willard-yuan Thank you for letting me know. I had seen that repository previously, but not close enough to see that he provides the VGG 19 pre-trained model alongside the project.
After opening this issue, I found out about @kastnerkyle's **sklearn-theano** which I have been using since then. Fortunately, there have been great progress since then regarding access to pre-trained models. At the moment, in addition to the sklearn-theano, Lasagne provides some of the most used Caffe pre-trained models. Also there is a (not-yet-merged) branch in Keras to add support for importing pre-trained Caffe models. However, due to recent substantial changes in Keras' API, I don't think it would be merged in the near future. 
It's also worth mentioning that there is [theano-alexnet](https://github.com/uoguelph-mlrg/theano_alexnet). It's a pure Theano implementation of the AlexNet model along with the weights. So one can also take a look at this, too. So far I can say that this is the only model created with Theano that has actually **been trained** on the ImageNet dataset.

**Comment by erfannoury** [comment]
IMO, at least from the point of view of a *user*, not a *core developer*, the tying together of model definition and model weights is where the hassle stems. So the difference in details of implementation shouldn't prevent a library from using the pre-trained weights of a model coded using another library; although I agree that there can't be a complete decoupling.
